src/gui/widgets/FxLine.cpp

```diff
@@ -395,3 +415,57 @@ void FxLine::setStrokeInnerInactive( const QColor & c )
 {
 	m_strokeInnerInactive = c;
 }
+
+void FxLine::changeColor()
+{
+	QColorDialog colorDialog( Engine::fxMixer()->effectChannel( m_channelIndex )->m_color );
```

Comment:
Okay, so if we move this to ColorChooser, and assuming you want the track PR to use this too, this one will have to be merged before I can implement it there (or else duplicated changes, possible merge conflicts, etc.)

Reply:
Yes, I think that should be fine as this one shouldn't be too hard to get merged.

Reply:
ColorChooser doesn't have a .cpp as I mentioned below, and I moved all of these QColorDialogs to one for the entire class. I think this workaround won't be an issue?

Reply:
This will create one QColorDialog per FxLine (i.e. one per mixer channel) and keep it around. If you thought keeping a single list of colors in ColorChooser would "just take up memory", this is much worse.

I think the .cpp file belongs in src/gui/dialogs. If you're hesitant to create it because you're unsure how, I could probably do so locally and send a diff. If you're avoiding it for some other reason, please specify. I still think it's the best solution and would rather not merge a workaround to something that could be done the right way.

Reply:
Note also that initializing the dialog in FxLine still doesn't prevent code duplication when you later need to initialize it for tracks.

Reply:
Alright, what about a color dialog saved in the `FXMixer` class (the parent of all `FXLine`s)? Because if you want me to use `ColorChooser` in place of `ColorDialog`, `ColorChooser` inherits `ColorDialog` so there's no improvement in using that instead. Probably the only thing that would be useful is not having to initialise the colors by the algorithms, which are anyways different for the mixer and the tracks (mixer colors are algorithmically darker and have a different saturation difference iirc).

I could make the required .cpp, but I didn't understand what exactly you intend to do here (no offence I'm prolly just dumb).

Reply:
> I didn't understand what exactly you intend to do here

To be honest, I thought the colors used were identical for tracks and mixer channels, so I really wanted them to be in one place. I see now that the mixer channel colors are darker by 10 (but otherwise identical). Is this necessary?

As I understand it, the mixer colors will be brightened/darkened/etc. based on selection and muting and so on. In that case, couldn't both the track and mixers use the same palette in the color picker, since the visible color isn't going to exactly match the selection anyway? If we can unify the palettes, then it still makes sense to consolidate them in a single class.

If you think they should have separate palettes, then it's no longer as appealing to put them all in the same class. In that case, saving the color dialog in FXMixer sounds like a good solution (at least if you can find a similar solution for #5573).

> `ColorDialog`, `ColorChooser` inherits `ColorDialog` so there's no improvement in using that instead.

It might be a good idea anyway, as QColorDialog was extended for a reason:
`//Forward key events to the parent to prevent stuck notes when the dialog gets focus`

----

For clarity's sake, my original idea was to make the track color palette the default palette for ColorChooser. This could be a very minor addition to ColorChooser: [gist](https://gist.github.com/Spekular/02e4a8a414f1b0ed3be095df197a51f4)

If we want to support multiple palettes, the class grows a bit more: [gist](https://gist.github.com/Spekular/00e784ee0e15a5cf6d374b2ed676a108)

Reply:
I suppose if I play around with the logic I can get away with one algorithm (or I can use your second suggestion). I'll see what I can do, and then commit if I get something going.

However, in the end the color chooser will have to stay somewhere, so either we let it be redeclared each time we need the palette, or save it in FXMixer. I'm a fan of the latter, but if you have better suggestions, I'm all ears.